src/osd/scrubber/scrub_queue.h

```diff
@@ -0,0 +1,404 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include "osd/osd_types_fmt.h"
+#include "osd/scrubber/scrub_queue_if.h"
+#include "osd/scrubber/scrub_resources.h"
+
+namespace Scrub {
+class ScrubSchedListener;
+class ScrubJob;
+struct SchedEntry;
+}  // namespace Scrub
+
+// clang-format off
+/*
+ * The main components involved in managing a "scrub initiation loop"
+ * (the process, initiated from within the OSD tick, to find a PG to
+ * scrub - but one that actually manages to reserve its replicas and
+ * start scrubbing) are:
+ *
+ *  OSD
+ * |
+ * |
+ * |
+ * |
+ * |    ScrubQueue
+ * |    -----------
+ * |
+ * |           |
+ * |           |
+ * +-> initiate_a_scrub()
+ *             |
+ *             +--->  selects a PG
+ *             |
+ *             +--->  initializes m_initiation_loop (an object detailing the
+ *             |      state of the current loop, i.e. the current attempt to
+ *             |      find a PG to scrub)
+ *             |
+ *             |      +------------------------+
+ *             |      | - restrictions         |
+ *             |      | - max tries            |
+ *             |      | - started at           |
+ *             |      | - ....                 |
+ *             |      |                        |
+ *             |      |                        |
+ *             |      |                        |
+ *             |      +------------------------+
+ *             |                                 PG & PgScrubber
+ *             |                                ------------------
+ *             +-------------------------------> start_scrubbing()
+ *             |                                  +
+ *             |                                  start_scrub()
+ *             |                                  | |
+ *             |                                  | |
+ *             |                                  | |creates a wrapper object around
+ *             |                                  | |"I am the chosen one in the
+ *             |                                  | | current loop" (m_schedloop_step)
+ *             |                                  | |
+ *             |                                  | |
+ *             |                                  | |Guaranteed to cause (dtor) one of:
+ *             |        loop ID (matching ^)      | |
+ *             |  <-------------------------------+-+-- try the next one in queue
+ *             |        scrub_next_in_queue()     | |   (as I cannot be scrubbed)
+ *             |                                  | |
+ *             |                                  | |Or:
+ *             |                                  | |
+ *             |  <-------------------------------+-+-- loop concluded due to failure
+ *             |                                  | |
+ *             |        initiation_loop_done()    | |Or:
```

Comment:
This choice of interface structure exposes a lot of the scrub queue details to the PG.  I don't really understand why we can't signal everything we need to signal with an asynchronous event to start scrubbing and an invocation of a scheduling interface for a new time (either because the PG was unable to start scrubbing, because the PG completed its scrub, or because the PG went through an interval change).